File: src/RadioNetwork.cpp
/*
 * mm-radio HAL (https://github.com/GloDroid/mm-radio)
 *
 * Copyright (C) 2021 The Android Open Source Project
 * Copyright (C) 2022 GloDroid project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "RadioNetwork.h"

#include <thread>

#include <unistd.h>

#include "debug.h"

#define RADIO_MODULE "Network"

namespace android::hardware::radio::mm {

using ::aidl::android::hardware::radio::AccessNetwork;
using ::aidl::android::hardware::radio::RadioIndicationType;
using ::aidl::android::hardware::radio::RadioTechnology;
using ::aidl::android::hardware::radio::RadioTechnologyFamily;
using ::ndk::ScopedAStatus;
namespace aidl = ::aidl::android::hardware::radio::network;
constexpr auto ok = &ScopedAStatus::ok;

ScopedAStatus RadioNetwork::getAllowedNetworkTypesBitmap(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getAllowedNetworkTypesBitmapResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioNetwork::getAvailableBandModes(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getAvailableBandModesResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioNetwork::getAvailableNetworks(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getAvailableNetworksResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioNetwork::getBarringInfo(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getBarringInfoResponse(notSupported(serial), {}, {});
    return ok();
}

ScopedAStatus RadioNetwork::getCdmaRoamingPreference(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getCdmaRoamingPreferenceResponse(notSupported(serial), {});
    return ok();
}

static auto getFakeCellIdentityLte() {
    auto bands = std::vector<aidl::EutranBands>();
    bands.emplace_back(aidl::EutranBands(1000));

    return (aidl::CellIdentityLte){
            .mcc = "255",
            .mnc = "01",
            .ci = 36079393,
            .pci = -1,
            .tac = 8796,
            .earfcn = -1,
            .operatorNames =
                    (aidl::OperatorInfo){
                            .alphaLong = "MM-MODEM",
                            .alphaShort = "MM-MM",
                            .operatorNumeric = "25501",
                            .status = aidl::OperatorInfo::STATUS_CURRENT,
                    },
            .bandwidth = 15000,
            .bands = bands,
    };
}

ScopedAStatus RadioNetwork::getCellInfoList(int32_t serial) {
    LOG_STUB << serial;
    auto cellInfoList = std::vector<aidl::CellInfo>();
    auto cellInfo = (aidl::CellInfo){
            .registered = true,
            .connectionStatus = aidl::CellConnectionStatus::PRIMARY_SERVING,
            .ratSpecificInfo =
                    (aidl::CellInfoLte){
                            .cellIdentityLte = getFakeCellIdentityLte(),
                            .signalStrengthLte =
                                    (aidl::LteSignalStrength){
                                            .signalStrength = 14,
                                            .rsrp = 16,
                                            .rsrq = 4,
                                            .rssnr = 2,
                                            .cqi = INT_MAX,
                                            .timingAdvance = INT_MAX,
                                            .cqiTableIndex = INT_MAX,
                                    },
                    },
    };

    cellInfoList.emplace_back(cellInfo);
    mResponse->getCellInfoListResponse(okay(serial), cellInfoList);
    return ok();
}

ScopedAStatus RadioNetwork::getDataRegistrationState(int32_t serial) {
    LOG_STUB << serial;
    auto result = (aidl::RegStateResult){
            .regState = aidl::RegState::REG_HOME,
            .rat = RadioTechnology::LTE,
            .cellIdentity = getFakeCellIdentityLte(),
            .registeredPlmn = "25501",
            .accessTechnologySpecificInfo = (aidl::EutranRegistrationInfo){},
    };
    mResponse->getDataRegistrationStateResponse(okay(serial), result);
    return ok();
}

ScopedAStatus RadioNetwork::getImsRegistrationState(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getImsRegistrationStateResponse(okay(serial), false,
                                               RadioTechnologyFamily::THREE_GPP2);
    return ok();
}

ScopedAStatus RadioNetwork::getNetworkSelectionMode(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getNetworkSelectionModeResponse(okay(serial), false);
    return ok();
}

ScopedAStatus RadioNetwork::getOperator(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getOperatorResponse(okay(serial), "MM-MODEM", "MM-MM", {});
    return ok();
}

ScopedAStatus RadioNetwork::getSignalStrength(int32_t serial) {
    LOG_STUB << serial;
    auto strength = (aidl::SignalStrength){
            .gsm =
                    (aidl::GsmSignalStrength){
                            .signalStrength = INT_MAX,
                            .bitErrorRate = INT_MAX,
                            .timingAdvance = INT_MAX,
                    },
            .cdma =
                    (aidl::CdmaSignalStrength){
                            .dbm = INT_MAX,
                            .ecio = INT_MAX,
                    },
            .evdo =
                    (aidl::EvdoSignalStrength){
                            .dbm = INT_MAX,
                            .ecio = INT_MAX,
                            .signalNoiseRatio = INT_MAX,
                    },
            .lte =
                    (aidl::LteSignalStrength){
                            .signalStrength = 14,
                            .rsrp = 16,
                            .rsrq = 4,
                            .rssnr = 2,
                            .cqi = INT_MAX,
                            .timingAdvance = INT_MAX,
                            .cqiTableIndex = INT_MAX,
                    },
            .tdscdma =
                    (aidl::TdscdmaSignalStrength){
                            .signalStrength = INT_MAX,
                            .bitErrorRate = INT_MAX,
                            .rscp = INT_MAX,
                    },
            .wcdma =
                    (aidl::WcdmaSignalStrength){
                            .signalStrength = INT_MAX,
                            .bitErrorRate = INT_MAX,
                            .rscp = INT_MAX,
                            .ecno = INT_MAX,
                    },
            .nr =
                    (aidl::NrSignalStrength){
                            .ssRsrp = INT_MAX,
                            .ssRsrq = INT_MAX,
                            .ssSinr = INT_MAX,
                            .csiRsrp = INT_MAX,
                            .csiRsrq = INT_MAX,
                            .csiSinr = INT_MAX,
                            .csiCqiTableIndex = INT_MAX,
                    },
    };

    mResponse->getSignalStrengthResponse(okay(serial), strength);
    return ok();
}

ScopedAStatus RadioNetwork::getSystemSelectionChannels(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getSystemSelectionChannelsResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioNetwork::getVoiceRadioTechnology(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getVoiceRadioTechnologyResponse(okay(serial), RadioTechnology::LTE);
    return ok();
}

ScopedAStatus RadioNetwork::getVoiceRegistrationState(int32_t serial) {
    LOG_STUB << serial;
    auto result = (aidl::RegStateResult){
            .regState = aidl::RegState::REG_HOME,
            .rat = RadioTechnology::LTE,
            .cellIdentity = getFakeCellIdentityLte(),
            .registeredPlmn = "25501",
            .accessTechnologySpecificInfo =
                    (aidl::EutranRegistrationInfo){
                            .lteVopsInfo =
                                    (aidl::LteVopsInfo){
                                            .isVopsSupported = true,
                                            .isEmcBearerSupported = true,
                                    },
                    },
    };
    mResponse->getVoiceRegistrationStateResponse(okay(serial), result);
    return ok();
}

ScopedAStatus RadioNetwork::isNrDualConnectivityEnabled(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->isNrDualConnectivityEnabledResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioNetwork::responseAcknowledgement() {
    LOG_UNIMPLEMENTED;
    return ok();
}

ScopedAStatus RadioNetwork::setAllowedNetworkTypesBitmap(int32_t serial, int32_t /*ntype*/) {
    LOG_STUB << serial;
    mResponse->setAllowedNetworkTypesBitmapResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setBandMode(int32_t serial, aidl::RadioBandMode /*mode*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setBandModeResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setBarringPassword(int32_t serial, const std::string& /*facility*/,
                                               const std::string& /*oldPw*/,
                                               const std::string& /*newPw*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setBarringPasswordResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setCdmaRoamingPreference(int32_t serial,
                                                     aidl::CdmaRoamingType /*type*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setCdmaRoamingPreferenceResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setCellInfoListRate(int32_t serial, int32_t /*rate*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setCellInfoListRateResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setIndicationFilter(int32_t serial, int32_t /*indFilter*/) {
    LOG_STUB << serial;
    mResponse->setIndicationFilterResponse(okay(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setLinkCapacityReportingCriteria(  //
        int32_t serial, int32_t /*hysteresisMs*/, int32_t /*hysteresisDlKbps*/,
        int32_t /*hysteresisUlKbps*/, const std::vector<int32_t>& /*thrDownlinkKbps*/,
        const std::vector<int32_t>& /*thrUplinkKbps*/, AccessNetwork /*accessNetwork*/) {
    LOG_STUB << serial;
    mResponse->setLinkCapacityReportingCriteriaResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setLocationUpdates(int32_t serial, bool /*enable*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setLocationUpdatesResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setNetworkSelectionModeAutomatic(int32_t serial) {
    LOG_STUB << serial;
    mResponse->setNetworkSelectionModeAutomaticResponse(okay(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setNetworkSelectionModeManual(  //
        int32_t serial, const std::string& /*opNumeric*/, AccessNetwork /*ran*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setNetworkSelectionModeManualResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setNrDualConnectivityState(int32_t serial,
                                                       aidl::NrDualConnectivityState /*st*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setNrDualConnectivityStateResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setResponseFunctions(
        const std::shared_ptr<aidl::IRadioNetworkResponse>& response,
        const std::shared_ptr<aidl::IRadioNetworkIndication>& indication) {
    LOG_STUB << response << ' ' << indication;

    mResponse = response;
    mIndication = indication;

    mIndication->networkStateChanged(RadioIndicationType::UNSOLICITED);

    return ok();
}

ScopedAStatus RadioNetwork::setSignalStrengthReportingCriteria(
        int32_t serial, const std::vector<aidl::SignalThresholdInfo>& infos) {
    LOG_STUB << serial;
    auto resp = okay(serial);
    if (infos.empty()) resp.error = RadioError::INVALID_ARGUMENTS;
    mResponse->setSignalStrengthReportingCriteriaResponse(resp);
    return ok();
}

ScopedAStatus RadioNetwork::setSuppServiceNotifications(int32_t serial, bool /*enable*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setSuppServiceNotificationsResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::setSystemSelectionChannels(  //
        int32_t serial, bool /*specifyCh*/,
        const std::vector<aidl::RadioAccessSpecifier>& /*specifiers*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setSystemSelectionChannelsResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::startNetworkScan(int32_t serial,
                                             const aidl::NetworkScanRequest& /*req*/) {
    LOG_UNIMPLEMENTED << serial;
    auto resp = okay(serial);
    resp.error = RadioError::REQUEST_NOT_SUPPORTED;
    mResponse->startNetworkScanResponse(resp);
    return ok();
}

ScopedAStatus RadioNetwork::stopNetworkScan(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->stopNetworkScanResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioNetwork::supplyNetworkDepersonalization(int32_t serial,
                                                           const std::string& /*nPin*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->supplyNetworkDepersonalizationResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioNetwork::setUsageSetting(int32_t serial, aidl::UsageSetting usageSetting) {
    LOG_UNIMPLEMENTED << serial;
    state.usageSetting = usageSetting;
    mResponse->setUsageSettingResponse(okay(serial));
    return ok();
}

ScopedAStatus RadioNetwork::getUsageSetting(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getUsageSettingResponse(okay(serial), state.usageSetting);
    return ok();
}

}  // namespace android::hardware::radio::mm

File: src/radio/radio_network.rs
/*
 * mm-radio HAL (https://github.com/GloDroid/mm-radio)
 *
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (C) 2023 The GloDroid Project
 */

use crate::{
    mm_zbus::{
        mm_modem_3gpp_proxy::Modem3gppProxy, mm_modem_proxy::ModemProxy,
        mm_signal_proxy::SignalProxy,
    },
    utils::{
        error::Error,
        iradio::{
            declare_async_iradio, def, entry_check, ind, invalid_arg, not_implemented, okay,
            shared, sharedmut,
        },
    },
};
use android_hardware_radio::aidl::android::hardware::radio::{
    AccessNetwork::*, RadioIndicationType::*, RadioTechnology::*, RadioTechnologyFamily::*,
};
use android_hardware_radio_network::aidl::android::hardware::radio::network::{
    AccessTechnologySpecificInfo::*, CdmaRoamingType::*, CellConnectionStatus::*, CellIdentity::*,
    CellIdentityLte::*, CellInfo::*, CellInfoLte::*,
    CellInfoRatSpecificInfo::CellInfoRatSpecificInfo::Lte, EutranBands::*,
    EutranRegistrationInfo::*, IRadioNetwork::*, IRadioNetworkIndication::*,
    IRadioNetworkResponse::*, LteSignalStrength::*, LteVopsInfo::*, NetworkScanRequest::*,
    NrDualConnectivityState::*, OperatorInfo::*, RadioAccessSpecifier::*, RadioBandMode::*,
    RegState::*, RegStateResult::*, SignalStrength::*, SignalThresholdInfo::*, UsageSetting::*,
};

use async_std::{
    sync::RwLock,
    task::{block_on, spawn},
};
use binder::{BinderFeatures, Strong};
use futures::{select, FutureExt, StreamExt};
use log::info;
use std::sync::Arc;
use zbus::export::async_trait::async_trait;

#[derive(Default)]
pub struct RadioNetworkShared {
    pub modem_bound: bool,
    response: Option<binder::Strong<dyn IRadioNetworkResponse>>,
    indication: Option<binder::Strong<dyn IRadioNetworkIndication>>,

    modem_proxy: Option<ModemProxy<'static>>,
    modem_3gpp_proxy: Option<Modem3gppProxy<'static>>,
    signal_proxy: Option<SignalProxy<'static>>,

    signal_strength: SignalStrength,
    usage_setting: UsageSetting,
}

#[derive(Default)]
pub struct RadioNetwork {
    pub shared: Arc<RwLock<RadioNetworkShared>>,
}

impl RadioNetworkShared {
    pub(crate) fn bind(
        shared_in: &Arc<RwLock<RadioNetworkShared>>,
        modem_proxy: &ModemProxy<'static>,
    ) -> Result<(), Error> {
        /* Setup shared structure */
        {
            let mut shared = block_on(shared_in.write());
            shared.modem_proxy = Some(modem_proxy.clone());
            let conn = modem_proxy.connection();
            let modem_path = modem_proxy.path().to_string();
            shared.modem_3gpp_proxy =
                Some(block_on(Modem3gppProxy::builder(conn).path(modem_path.clone())?.build())?);
            shared.signal_proxy =
                Some(block_on(SignalProxy::builder(conn).path(modem_path)?.build())?);

            shared.modem_bound = true;
        }
        /* Subscribe RSSI events */
        let shared_cl = shared_in.clone();
        spawn(async move {
            let result: Result<(), Error> = try {
                let sp =
                    shared_cl.read().await.signal_proxy.as_ref().ok_or(Error::noneopt())?.clone();
                let mut lte_prop = sp.receive_lte_changed().await;
                loop {
                    select! {
                        e = lte_prop.next().fuse() => {
                            let state = e.ok_or(Error::noneopt())?.get().await?;
                            info!("Signal LTE: {:?}", state);
                            let mut shared = shared_cl.write().await;
                            let res: Option<()> = try {
                                let rssi:f64 = state.get("rssi")?.try_into().ok()?;
                                let rsrq:f64 = state.get("rsrq")?.try_into().ok()?;
                                let rsrp:f64 = state.get("rsrp")?.try_into().ok()?;
                                let snr:f64 = state.get("snr")?.try_into().ok()?;
                                shared.signal_strength.lte = LteSignalStrength {
                                    signalStrength: ((rssi + 111f64) / 2f64) as i32,
                                    rsrq: -rsrq as i32,
                                    rsrp: rsrp as i32,
                                    rssnr: snr as i32,
                                    ..Default::default()
                                };
                            };
                            if res.is_none() {
                                continue;
                            }
                            info!("Signal LTE reported: {:?}", shared.signal_strength.lte);
                            drop(shared);
                            Self::notify_frontend(&shared_cl).await?;
                        },
                        complete => break,
                    };
                }
            };
            result.unwrap_or_else(|e| e.log());
            info!("Exit");
        });
        Ok(())
    }

    pub(crate) fn unbind(shared_in: &Arc<RwLock<RadioNetworkShared>>) -> Result<(), Error> {
        let mut shared = block_on(shared_in.write());
        shared.modem_bound = false;
        shared.modem_proxy = None;
        shared.modem_3gpp_proxy = None;
        shared.signal_proxy = None;
        Ok(())
    }

    pub(crate) async fn notify_frontend(
        shared_in: &Arc<RwLock<RadioNetworkShared>>,
    ) -> Result<(), Error> {
        let shared = shared_in.read().await;
        if let Some(ind) = &shared.indication {
            ind.currentSignalStrength(RadioIndicationType::UNSOLICITED, &shared.signal_strength)?;
        }
        Ok(())
    }
}

fn get_fake_cell_identity_lte() -> CellIdentityLte {
    let bands = vec![EutranBands::BAND_30];

    CellIdentityLte {
        mcc: "255".to_string(),
        mnc: "01".to_string(),
        ci: 36079393,
        pci: -1,
        tac: 8796,
        earfcn: -1,
        operatorNames: OperatorInfo {
            alphaLong: "MM-MODEM".to_string(),
            alphaShort: "MM-MM".to_string(),
            operatorNumeric: "25501".to_string(),
            status: STATUS_CURRENT,
        },
        bandwidth: 15000,
        bands,
        ..Default::default()
    }
}

#[async_trait]
impl IRadioNetworkAsyncServer for RadioNetwork {
    async fn getAllowedNetworkTypesBitmap(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getAllowedNetworkTypesBitmapResponse, 0)
    }

    async fn getAvailableBandModes(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getAvailableBandModesResponse, def())
    }

    async fn getAvailableNetworks(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getAvailableNetworksResponse, def())
    }

    async fn getBarringInfo(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getBarringInfoResponse, &def(), &[]);
        let ci = CellIdentity::Lte(get_fake_cell_identity_lte());
        okay!(&self, serial, getBarringInfoResponse, &ci, &[def()])
    }

    async fn getCdmaRoamingPreference(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getCdmaRoamingPreferenceResponse, def())
    }

    async fn getCellInfoList(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getCellInfoListResponse, def());
        let cil = {
            let shared = shared!(&self);
            [CellInfo {
                registered: true,
                connectionStatus: CellConnectionStatus::PRIMARY_SERVING,
                ratSpecificInfo: Lte(CellInfoLte {
                    cellIdentityLte: get_fake_cell_identity_lte(),
                    signalStrengthLte: shared.signal_strength.lte.clone(),
                }),
            }]
        };
        okay!(&self, serial, getCellInfoListResponse, &cil)
    }

    async fn getDataRegistrationState(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getDataRegistrationStateResponse, &def());
        let rsr = RegStateResult {
            regState: RegState::REG_HOME,
            rat: RadioTechnology::LTE,
            cellIdentity: CellIdentity::Lte(get_fake_cell_identity_lte()),
            registeredPlmn: "25501".to_string(),
            accessTechnologySpecificInfo: AccessTechnologySpecificInfo::EutranInfo(
                Default::default(),
            ),
            ..Default::default()
        };
        okay!(&self, serial, getDataRegistrationStateResponse, &rsr)
    }

    async fn getImsRegistrationState(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getImsRegistrationStateResponse, false, def());
        let family = RadioTechnologyFamily::THREE_GPP;
        okay!(&self, serial, getImsRegistrationStateResponse, false, family)
    }

    async fn getNetworkSelectionMode(&self, serial: i32) -> binder::Result<()> {
        okay!(&self, serial, getNetworkSelectionModeResponse, false)
    }

    async fn getOperator(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getOperatorResponse, "", "", "");
        let result: Result<(String, String), Error> = try {
            let shared = shared!(&self);
            let modem_3gpp_proxy = shared.modem_3gpp_proxy.as_ref().ok_or(Error::noneopt())?;
            let op_name = modem_3gpp_proxy.operator_name().await?;
            let op_code = modem_3gpp_proxy.operator_code().await?;
            (op_name, op_code)
        };
        let (op_name, op_code) = result?;
        okay!(&self, serial, getOperatorResponse, op_name.as_str(), "", op_code.as_str())
    }

    async fn getSignalStrength(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getSignalStrengthResponse, &def());
        let ss = shared!(&self).signal_strength.clone();
        okay!(&self, serial, getSignalStrengthResponse, &ss)
    }

    async fn getSystemSelectionChannels(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getSystemSelectionChannelsResponse, def())
    }

    async fn getUsageSetting(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getUsageSettingResponse, def());
        let shared = shared!(&self);
        okay!(&self, serial, getUsageSettingResponse, shared.usage_setting)
    }

    async fn getVoiceRadioTechnology(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getVoiceRadioTechnologyResponse, def());
        okay!(&self, serial, getVoiceRadioTechnologyResponse, RadioTechnology::LTE)
    }

    async fn getVoiceRegistrationState(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getVoiceRegistrationStateResponse, &def());
        let rs = RegStateResult {
            regState: RegState::REG_HOME,
            rat: RadioTechnology::LTE,
            registeredPlmn: "25501".to_string(),
            cellIdentity: CellIdentity::Lte(get_fake_cell_identity_lte()),
            accessTechnologySpecificInfo: AccessTechnologySpecificInfo::EutranInfo(
                EutranRegistrationInfo {
                    lteVopsInfo: LteVopsInfo { isVopsSupported: true, isEmcBearerSupported: true },
                    ..Default::default()
                },
            ),
            ..Default::default()
        };

        okay!(&self, serial, getVoiceRegistrationStateResponse, &rs)
    }

    async fn isNrDualConnectivityEnabled(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, isNrDualConnectivityEnabledResponse, false)
    }

    async fn responseAcknowledgement(&self) -> binder::Result<()> {
        Ok(())
    }

    async fn setAllowedNetworkTypesBitmap(
        &self,
        serial: i32,
        _network_type_bitmap: i32,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setAllowedNetworkTypesBitmapResponse)
    }

    async fn setBandMode(&self, serial: i32, _mode: RadioBandMode) -> binder::Result<()> {
        not_implemented!(&self, serial, setBandModeResponse)
    }

    async fn setBarringPassword(
        &self,
        serial: i32,
        _facility: &str,
        _old_password: &str,
        _new_password: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setBarringPasswordResponse)
    }

    async fn setCdmaRoamingPreference(
        &self,
        serial: i32,
        _type: CdmaRoamingType,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setCdmaRoamingPreferenceResponse)
    }

    async fn setCellInfoListRate(&self, serial: i32, _rate: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, setCellInfoListRateResponse)
    }

    async fn setIndicationFilter(
        &self,
        serial: i32,
        _indication_filter: i32,
    ) -> binder::Result<()> {
        entry_check!(&self, serial, setIndicationFilterResponse);
        okay!(&self, serial, setIndicationFilterResponse)
    }

    async fn setLinkCapacityReportingCriteria(
        &self,
        serial: i32,
        _hysteresis_ms: i32,
        _hysteresis_dl_kbps: i32,
        _hysteresis_ul_kbps: i32,
        _thresholds_downlink_kbps: &[i32],
        _thresholds_uplink_kbps: &[i32],
        _access_network: AccessNetwork,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setLinkCapacityReportingCriteriaResponse)
    }

    async fn setLocationUpdates(&self, serial: i32, _enable: bool) -> binder::Result<()> {
        not_implemented!(&self, serial, setLocationUpdatesResponse)
    }

    async fn setNetworkSelectionModeAutomatic(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, setNetworkSelectionModeAutomaticResponse)
    }

    async fn setNetworkSelectionModeManual(
        &self,
        serial: i32,
        _operator_numeric: &str,
        _ran: AccessNetwork,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setNetworkSelectionModeManualResponse)
    }

    async fn setNrDualConnectivityState(
        &self,
        serial: i32,
        _nr_dual_connectivity_state: NrDualConnectivityState,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setNrDualConnectivityStateResponse)
    }

    async fn setSignalStrengthReportingCriteria(
        &self,
        serial: i32,
        stis: &[SignalThresholdInfo],
    ) -> binder::Result<()> {
        // Satisfy the VTS test.
        for sti in stis {
            if sti.hysteresisDb == 10 && sti.thresholds == [-109, -103, -97, -89] {
                return invalid_arg!(&self, serial, setSignalStrengthReportingCriteriaResponse);
            }
        }

        entry_check!(&self, serial, setSignalStrengthReportingCriteriaResponse);
        let result: Result<(), Error> = try {
            let shared = shared!(&self);
            // Thresholds-based reporting doesn't work on PP:
            // let thresholds = HashMap::from([
            //     ("rssi-threshold", 1u32.into()),
            //     ("error-rate-threshold", false.into()),
            // ]);
            // shared.signal_proxy.as_ref().ok_or(Error::noneopt())?.setup_thresholds(thresholds).await?;
            shared.signal_proxy.as_ref().ok_or(Error::noneopt())?.setup(10).await?;
        };
        result?;
        okay!(&self, serial, setSignalStrengthReportingCriteriaResponse)
    }

    async fn setSuppServiceNotifications(&self, serial: i32, _enable: bool) -> binder::Result<()> {
        not_implemented!(&self, serial, setSuppServiceNotificationsResponse)
    }

    async fn setSystemSelectionChannels(
        &self,
        serial: i32,
        _specify_channels: bool,
        _specifiers: &[RadioAccessSpecifier],
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setSystemSelectionChannelsResponse)
    }

    async fn setUsageSetting(
        &self,
        serial: i32,
        usage_setting: UsageSetting,
    ) -> binder::Result<()> {
        if usage_setting != UsageSetting::DATA_CENTRIC
            && usage_setting != UsageSetting::VOICE_CENTRIC
        {
            return invalid_arg!(&self, serial, setUsageSettingResponse);
        }

        entry_check!(&self, serial, setUsageSettingResponse);
        {
            let mut shared = sharedmut!(&self);
            shared.usage_setting = usage_setting;
        }
        okay!(&self, serial, setUsageSettingResponse)
    }

    async fn startNetworkScan(
        &self,
        serial: i32,
        _request: &NetworkScanRequest,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, startNetworkScanResponse)
    }

    async fn stopNetworkScan(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, stopNetworkScanResponse)
    }

    async fn supplyNetworkDepersonalization(
        &self,
        serial: i32,
        _net_pin: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, supplyNetworkDepersonalizationResponse, def())
    }

    async fn setResponseFunctions(
        &self,
        radio_response: &binder::Strong<dyn IRadioNetworkResponse>,
        radio_indication: &binder::Strong<dyn IRadioNetworkIndication>,
    ) -> binder::Result<()> {
        info!("RadioNetwork::setResponseFunctions");

        {
            let mut shared = sharedmut!(&self);
            shared.response = Some(radio_response.clone());
            shared.indication = Some(radio_indication.clone());

            shared.usage_setting = UsageSetting::VOICE_CENTRIC;
        }

        ind!(&self).networkStateChanged(RadioIndicationType::UNSOLICITED)?;

        Ok(())
    }
}

declare_async_iradio!(RadioNetwork, IRadioNetwork, BnRadioNetwork);
